### PR TITLE
Rewrite command and overload builders using C#7 features

### PR DESCRIPTION
**Pull request by FiniteReality**
Passes simple test cases and some more complicated ones, I would appreciate more testers though.

EDIT: I forgot to mention that this is *definitely* a breaking change.

### CODE REVIEW

File: src/Discord.Net.Commands/Builders/CommandBuilder.cs [hunk 1]
```diff
@@ -8,43 +8,41 @@ namespace Discord.Commands.Builders
...
             Module = module;
 
-            _preconditions = new List<PreconditionAttribute>();
-            _parameters = new List<ParameterBuilder>();
+            _overloads = new List<OverloadBuilder>();
             _aliases = new List<string>();
         }
         //User-defined
-        internal CommandBuilder(ModuleBuilder module, string primaryAlias, Func<ICommandContext, object[], IServiceProvider, Task> callback)
+        internal CommandBuilder(ModuleBuilder module, string primaryAlias, Action<OverloadBuilder> defaultOverloadBuilder)
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/Builders/CommandBuilder.cs:33]: Why go from a `Func<..., Task>` to an `Action<...>`? Just thinking out loud.

File: src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs [hunk 2]
```diff
@@ -196,23 +229,29 @@ private static void BuildParameter(ParameterBuilder builder, System.Reflection.P
             foreach (var attribute in attributes)
             {
                 // TODO: C#7 type switch
```
Comment (Auralytical) [inline src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs]: You left a TODO in

File: src/Discord.Net.Commands/Builders/CommandBuilder.cs (hunk 1 above)
Comment (FiniteReality) [reply src/Discord.Net.Commands/Builders/CommandBuilder.cs:33]: The overload builder contains the callback, so it'd be something like this:

```
module.AddCommand("name", overload => {
    overload.WithCallback(/*callback here*/);
});
```

File: src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs (hunk 2 above)
Comment (FiniteReality) [reply src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs]: Whoops. At least it was a comment and not missing code :stuck_out_tongue:

**Comment by FiniteReality** [comment]
After a discussion, we've decided it would be better to find another way to do this.

**Pull request closed without merging**